### ISSUE TITLE
Development docs unclear on adding new features

### ISSUE THREAD

**Issue by das-intensity** [dariusbakunas/truenas-go-sdk#14]
I was hoping to extend this and add a few extra things like `/interface`, but I just want to be sure I understand the dev flow.

From what I can tell:
1. `cfg/bluefin_original.yaml` is the automatically-generated docs as per fetch.sh against an up-to-date bluefin instance
2. `cfg/bluefin.yaml` is **manually** written, using `bluefin_original.yaml` only as a general reference
3. The reason to do this is that the truenas live api docs aren't compatible with openapi generator
4. Testing is done manually via programs that use the sdk (like `terraform-provider-truenas`)

Is that about correct?

**Comment by dariusbakunas** [comment]
Yep, and `generate.sh` script will run openapi sdk for you with all parameters set, it takes `release` name as first argument and currently only bluefin and angelfish supported. Original spec is not only not compatible, but it is also missing a lot of response/input type definitions.

**Comment by das-intensity** [comment]
Any idea where we'd need to edit (upstream) in order to have some of these corrected in the actual `/api/v2.0` page? It'd be nice if this could (one day) be reduced to:

* A curated list of routes whose upstream API docs are correct
* A curated set of routes/schemas whose upstream API docs are NOT correct (but that we still want in the sdk)
* A translation layer to convert some of the incompatibilities (e.g. openapi generator doesn't like the `_name_` field in schemas)

Then the generate script could:

* Run the translation on the upstream API yaml
* Copy over the "known good" routes from translated upstream API yaml
* Merge in the extra ones that aren't valid upstream

Looking at the differences between `bluefin.yaml` and `bluefin_original.yaml` for some routes, it seems like a heavy lift, but not necessarily unachievable (assuming truenas upstream is willing to accept the changes needed on their side). What do you think? Any wisdom on where such an effort is likely to fall flat?

**Comment by dariusbakunas** [comment]
Yes, that would be ideal. Last time I needed something fixed on their end, I posted in their forums - maybe thats the way to go. That would definitely make my life easier. I just recently started adding bluefin support and it is a sizable undertaking, honestly making me lose faith in this project :) I can't imagine adding yet another TrueNAS release support to this.

**Comment by das-intensity** [comment]
Good news on the upstream front. Turns out the lack of schema on most routes was a bug introduced in 2021. I've put up a PR here: https://github.com/truenas/middleware/pull/10747, and since it's clear to see it wasn't an intentional change, I'm hopeful they'll accept it soon :)

Obviously not files you'd normally want to mess with, but if you want the updated schema in your docs, you can make the simple edit from the PR to this file on your trunas instance:
```
/usr/lib/python3/dist-packages/middlewared/restful.py
```

I guess the next big target is the JSON returned upon success for GET routes? They do have a `returns()` decorator similar to the `accepts()` one that generates the PUT schemas, so maybe it's only a minor change also? One can dream ;)

**Comment by dariusbakunas** [comment]
Wow, nice!
